### PR TITLE
Add ability to specify protocol a leaf device can use to connect to the Edge

### PR DESCRIPTION
**Pull request by mrohera**
* Add ability to specify a device protocol (MQTT or AMQP) a leaf device can use to connect to the Edge
* Fix to edge scope handling

### CODE REVIEW

File: smoke/LeafDevice/details/Details.cs [hunk 1]
```diff
@@ -148,8 +163,9 @@ protected async Task GetOrCreateDeviceIdentityAsync()
             IotHubConnectionStringBuilder builder = IotHubConnectionStringBuilder.Create(this.iothubConnectionString);
             RegistryManager rm = RegistryManager.CreateFromConnectionString(builder.ToString());
 
-            var edgeScope = Option.None<string>();
-            this.edgeDeviceId.ForEach(async (id) => edgeScope = await GetScopeIfExitsAsync(rm, id));
+            Option<string> edgeScope = await this.edgeDeviceId.Map(async (id) => {
+                return await GetScopeIfExitsAsync(rm, id);
+            }).GetOrElse(() => Task.FromResult<Option<string>>(Option.None<string>()));
```
Comment (myagley) [inline smoke/LeafDevice/details/Details.cs]: You don't need the `async`/`await` here. 

```csharp
Option<string> edgeScope = await this.edgeDeviceId
    .Map(id => GetScopeIfExistsAsync(rm, id))
    .GetOrElse(() => Task.FromResult<Option<string>>(Option.None<string>()));
```

File: smoke/LeafDevice/details/Details.cs [hunk 2]
```diff
@@ -49,6 +49,7 @@ public class Details
             string trustedCACertificateFileName,
             string edgeHostName,
             string edgeDeviceId,
+            string protocol,
```
Comment (varunpuranik) [inline smoke/LeafDevice/details/Details.cs]: If we are adding protocol, we can remove useWebSockets. Protocol values can be - 
Mqtt, MqttWs, Amqp and AmqpWs.. just like for the product.

File: smoke/LeafDevice/Program.cs [hunk 3]
```diff
@@ -53,6 +53,9 @@ class Program
         [Option("-ed-id|--edge-device-id", Description = "Device Id of the Edge device that acts as a gateway to the leaf device. If not provided, the leaf device will not be in the Edge device's scope")]
         public string EdgeGatewayDeviceId { get; } = string.Empty;
 
+        [Option("-proto|--protocol", Description = "Protocol the leaf device will use to communicate with the Edge device. Choices are MQTT and AMQP. If protocol is unspecified, default is MQTT.")]
+        public string protocol { get; } = "MQTT";
```
Comment (philipktlin) [inline smoke/LeafDevice/Program.cs]: can this be Enum instead of string?

Comment (philipktlin) [reply smoke/LeafDevice/Program.cs]: it is better to combine UseWebSockets parameter.

File: smoke/LeafDevice/details/Details.cs (hunk 1 above)
Comment (mrohera) [reply smoke/LeafDevice/details/Details.cs]: Thanks fixed.

File: smoke/LeafDevice/details/Details.cs [hunk 4]
```diff
@@ -25,6 +25,13 @@ namespace LeafDeviceTest
     using Message = Microsoft.Azure.Devices.Client.Message;
     using ServiceClientTransportType = Microsoft.Azure.Devices.TransportType;
 
+   public enum DeviceProtocol {
```
Comment (philipktlin) [inline smoke/LeafDevice/details/Details.cs]: should we use Microsoft.Azure.Devices.Client.TransportType?

Comment (mrohera) [reply smoke/LeafDevice/details/Details.cs]: That has more options than a user would want to set for our tests for example Http so it becomes a bit clunky.

**Pull request merged**